lib/irq.c

```diff
@@ -228,162 +235,122 @@ dev_set_irqs_validate(vfu_ctx_t *vfu_ctx, struct vfio_irq_set *irq_set)
+            irq_set->index, irq_set->start, irq_set->count, nr_fds);
+    return -EINVAL;
 }
 
-static long
-dev_set_irqs(vfu_ctx_t *vfu_ctx, struct vfio_irq_set *irq_set, void *data)
+int
+handle_device_set_irqs(vfu_ctx_t *vfu_ctx, uint32_t size,
+                       int *fds, size_t nr_fds, struct vfio_irq_set *irq_set)
 {
-    long ret;
+    uint32_t data_type;
+    int ret;
 
     assert(vfu_ctx != NULL);
     assert(irq_set != NULL);
 
-    // Ensure irq_set is valid.
-    ret = dev_set_irqs_validate(vfu_ctx, irq_set);
+    if (size < sizeof(*irq_set) || size != irq_set->argsz) {
+        vfu_log(vfu_ctx, LOG_ERR, "%s: bad size %u", __func__, size);
+        return -EINVAL;
+    }
+
+    ret = device_set_irqs_validate(vfu_ctx, irq_set, nr_fds);
     if (ret != 0) {
         return ret;
     }
 
     switch (irq_set->flags & VFIO_IRQ_SET_ACTION_TYPE_MASK) {
-    case VFIO_IRQ_SET_ACTION_MASK:     // fallthrough
+    case VFIO_IRQ_SET_ACTION_MASK:
     case VFIO_IRQ_SET_ACTION_UNMASK:
         // We're always edge-triggered without un/mask support.
+        // FIXME: return an error? We don't report MASKABLE
         return 0;
+    case VFIO_IRQ_SET_ACTION_TRIGGER:
+        break;
     }
 
-    return irqs_trigger(vfu_ctx, irq_set, data);
-}
-
-static long
-dev_get_irqinfo(vfu_ctx_t *vfu_ctx, struct vfio_irq_info *irq_info_in,
-                struct vfio_irq_info *irq_info_out)
-{
-    assert(vfu_ctx != NULL);
-    assert(irq_info_in != NULL);
-    assert(irq_info_out != NULL);
-
-    // Ensure provided argsz is sufficiently big and index is within bounds.
-    if ((irq_info_in->argsz < sizeof(struct vfio_irq_info)) ||
-        (irq_info_in->index >= VFU_DEV_NUM_IRQS)) {
-        vfu_log(vfu_ctx, LOG_DEBUG, "bad irq_info (size=%d index=%d)\n",
-                irq_info_in->argsz, irq_info_in->index);
-        return -EINVAL;
-    }
-
-    irq_info_out->count = vfu_ctx->irq_count[irq_info_in->index];
-    irq_info_out->flags = VFIO_IRQ_INFO_EVENTFD;
-
-    return 0;
-}
+    data_type = irq_set->flags & VFIO_IRQ_SET_DATA_TYPE_MASK;
 
-int
-handle_device_get_irq_info(vfu_ctx_t *vfu_ctx, uint32_t size,
-                           struct vfio_irq_info *irq_info_in,
-                           struct vfio_irq_info *irq_info_out)
-{
-    assert(vfu_ctx != NULL);
-    assert(irq_info_in != NULL);
-    assert(irq_info_out != NULL);
-
-    if (size != sizeof(*irq_info_in) || size != irq_info_in->argsz) {
-        vfu_log(vfu_ctx, LOG_WARNING, "IRQ info size %d", size);
-        return -EINVAL;
+    if ((data_type == VFIO_IRQ_SET_DATA_NONE && irq_set->count == 0) ||
+        (data_type == VFIO_IRQ_SET_DATA_EVENTFD && nr_fds == 0)) {
+        return irqs_disable(vfu_ctx, irq_set->index,
+                            irq_set->start, irq_set->count);
     }
 
-    return dev_get_irqinfo(vfu_ctx, irq_info_in, irq_info_out);
-}
+    vfu_log(vfu_ctx, LOG_DEBUG, "setting IRQ %s flags=%#x start=%u count=%u",
```

Comment:
IMO interval notation would be preferable.

Reply:
done